### PR TITLE
deprecate .h headers in favor of .hpp headers

### PR DESCRIPTION
**Pull request by mikaelarguedas**

### CODE REVIEW

File: include/class_loader/class_loader.h [hunk 1]
```diff
@@ -30,289 +32,10 @@
...
-  int load_ref_count_;
-  boost::recursive_mutex load_ref_count_mutex_;
-  int plugin_ref_count_;
-  boost::recursive_mutex plugin_ref_count_mutex_;
-  static bool has_unmananged_instance_been_created_;
-};
-
-}  // namespace class_loader
+// *INDENT-OFF* (prevent uncrustify from adding indention below)
+#warning Including header <class_loader/class_loader.h> is deprecated, \
```
Comment (dirk-thomas) [inline include/class_loader/class_loader.h:36]: I am not sure if this justifies a deprecation warning. Why can't existing code just continue to use the `.h` file without the need to change?

Comment (mikaelarguedas) [reply include/class_loader/class_loader.h:36]: The goal here is to remove these headers in a future version of ROS. In this version the users will be able to keep using the .h headers. The goal of the deprecation warning is to warn users to update their include statements to the new headers before they get removed (in the next ROS version). That's what we've been doing for pluginlib as well: https://github.com/ros/pluginlib/pull/70.

The motivation for changing the headers is to allow tools (like linters, IDE etc) to process them properly and to be more semantically correct. The new header layout will be backported to previous distributions and tickets will be filed on downstream packages to ease transition.

Comment (dirk-thomas) [reply include/class_loader/class_loader.h:36]: I just want to point out that you are "forcing" potentially a lot of maintainers to update their code for no "good" reason. I you are planning to make PR for each affected downstream package that would be good.

Comment (dirk-thomas) [reply include/class_loader/class_loader.h:36]: Just keep in mind that there might be more packages out there which might not be released or even public which all need to be updated.

Comment (mikaelarguedas) [reply include/class_loader/class_loader.h:36]: > I you are planning to make PR for each affected downstream package that would be good.

Yes that's the plan. I'm considering providing a script to convert as well for non public packages (ala https://github.com/ros/pluginlib/blob/kinetic-devel/scripts/plugin_macro_update).

> Just keep in mind that there might be more packages out there which might not be released or even public which all need to be updated.

Indeed, the "good" reason being: moving toward ROS1 / ROS2 compatibility. Arguably that's not a "good enough" reason, I'm mostly looking for a way to give users incentive to move forward without breaking their code. Apparently just mentioning it in the headers and documentation has not been enough in the past (in the case of pluginlib I found more than 70 packages still using the deprecated macros after several years).

Comment (mikaelarguedas) [reply include/class_loader/class_loader.h:36]: In conclusion I'll move forward with this with a few extra items:
- backport the new headers to all active ros distributions
- open PR on all downstream packages to use the new headers (once they're released on all distros)
- a script to perform the conversion automatically
- note about the new headers in the melodic migration guide

**Pull request merged**